File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/smt/SmtTermFactory.java
package com.github.jmlparser.smt;

import com.github.javaparser.resolution.types.ResolvedPrimitiveType;
import com.github.javaparser.resolution.types.ResolvedType;
import com.github.jmlparser.smt.model.SAtom;
import com.github.jmlparser.smt.model.SExpr;
import com.github.jmlparser.smt.model.SList;
import com.github.jmlparser.smt.model.SmtType;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import lombok.SneakyThrows;

import java.math.BigInteger;
import java.util.List;

/**
 * @author Alexander Weigl
 * @version 1 (07.08.22)
 */
public class SmtTermFactory {
    public static final SmtTermFactory INSTANCE = new SmtTermFactory();

    private final Cache<String, SAtom> symbolAndValueCache = CacheBuilder.newBuilder().softValues().build();

    //region boolean operators
    public SExpr and(SExpr... terms) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "and", terms);
    }

    public SExpr and(List<SExpr> seq) {
        return and(seq.toArray(new SExpr[0]));
    }

    public SExpr or(SExpr... terms) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "or", terms);
    }

    public SExpr impl(SExpr premise, SExpr concl) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "=>", premise, concl);

    }
    //endregion

    public SExpr ite(SExpr cond, SExpr then, SExpr otherwise) {
        return fnApply(then.getJavaType(), then.getSmtType(), "ite", cond, then, otherwise);
    }

    SExpr fnApply(ResolvedType javaType, SmtType smtType, String fn, SExpr arg) {
        return new SList(smtType, javaType, symbol(fn), arg);
    }

    private SExpr fnApply(ResolvedType javaType, SmtType smtType, String fn, SExpr arg1, SExpr arg2) {
        return new SList(smtType, javaType, symbol(fn), arg1, arg2);
    }

    private SExpr fnApply(ResolvedType javaType, SmtType smtType, String fn, SExpr arg1, SExpr arg2, SExpr arg3) {
        return new SList(smtType, javaType, symbol(fn), arg1, arg2, arg3);

    }

    private SExpr fnApply(ResolvedType javaType, SmtType smtType, String fn, SExpr... args) {
        return list(javaType, smtType, symbol(fn), args);
    }


    @SneakyThrows
    SAtom symbol(String fn) {
        return symbolAndValueCache.get(fn, () -> new SAtom(SmtType.SYMBOL, null, fn));
    }

    public SExpr forall(List<? extends SExpr> variables, SExpr formula) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "forall", list(variables), formula);
    }

    public SExpr exists(List<SExpr> variables, SExpr formula) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "exists", list(variables), formula);
    }


    SExpr list(ResolvedType javaType, SmtType smtType, SAtom symbol, SExpr[] args) {
        SExpr[] nargs = new SExpr[args.length + 1];
        nargs[0] = symbol;
        System.arraycopy(args, 0, nargs, 1, args.length);
        return new SList(smtType, javaType, nargs);
    }

    SExpr list(List<? extends SExpr> variables) {
        SExpr[] args = variables.toArray(new SExpr[0]);
        return new SList(null, null, args);
    }

    //region polymorphic operators
    public SExpr bor(SExpr left, SExpr right) {
        if (isBool(left, right)) return or(left, right);
        if (isBv(left, right)) return bvor(left, right);
        return typeException();
    }


    public SExpr add(SExpr left, SExpr right) {
        if (isInt(left, right)) return iadd(left, right);
        if (isBv(left, right)) return bvadd(left, right);
        return typeException();
    }

    public SExpr subtract(SExpr left, SExpr right) {
        if (isInt(left, right)) return isubstract(left, right);
        if (isBv(left, right)) return bvsubstract(left, right);
        return typeException();
    }

    public SExpr modulo(SExpr left, SExpr right, boolean b) {
        if (isInt(left, right)) return imodulo(left, right);
        if (isBv(left, right)) return bvmodulo(left, right);
        return typeException();
    }

    public SExpr shiftLeft(SExpr left, SExpr right) {
        if (isBv(left, right)) return bvshiftLeft(left, right);
        return typeException();
    }


    public SExpr shiftRight(SExpr left, SExpr right, boolean sign) {
        if (isBv(left, right)) return bvshiftRight(left, right, sign);
        return typeException();
    }

    public SExpr lessOrEquals(SExpr left, SExpr right, boolean sign) {
        if (isInt(left, right)) return ilessOrEquals(left, right);
        if (isBv(left, right)) return bvlessOrEquals(left, right);
        return typeException("Could not handle types '%s <= %s'", left.getSmtType(), right.getSmtType());
    }

    public SExpr greaterOrEquals(SExpr left, SExpr right, boolean b) {
        if (isInt(left, right)) return igreaterOrEquals(left, right);
        if (isBv(left, right)) return bvgreaterOrEquals(left, right);
        return typeException();
    }

    public SExpr lessThan(SExpr left, SExpr right) {
        if (isInt(left, right)) return ilessThan(left, right);
        if (isBv(left, right)) return bvlessThan(left, right);
        return typeException("Could not handle types '%s <%s'", left.getSmtType(), right.getSmtType());
    }

    public SExpr equiv(SExpr left, SExpr right) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "=", left, right);
    }

    public SExpr not(SExpr expr) {
        if (isBv(expr)) return bvnot(expr);
        if (isBool(expr)) return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "not", expr);
        return typeException();
    }


    public SExpr xor(SExpr left, SExpr right) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "xor", left, right);
    }

    public SExpr equality(SExpr left, SExpr right) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, "=", left, right);
    }

    public SExpr band(SExpr left, SExpr right) {
        if (isBool(left, right)) return and(left, right);

        if (isBv(left, right)) return bvand(left, right);

        return typeException();
    }


    private SExpr fnApplyToBool(String symbol, SExpr left, SExpr right) {
        return fnApply(ResolvedPrimitiveType.BOOLEAN, SmtType.BOOL, symbol, left, right);
    }

    public SExpr greaterThan(SExpr left, SExpr right) {
        if (isInt(left, right)) return igreaterThan(left, right);
        if (isBv(left, right)) return bvgreaterThan(left, right);
        return typeException("Could not handle types '%s > %s'", left.getSmtType(), right.getSmtType());
    }


    public SExpr negate(SExpr sexpr) {
        if (isBool(sexpr)) return not(sexpr);
        if (isBv(sexpr)) return bvnegate(sexpr);
        return typeException();
    }


    public SExpr multiply(SExpr left, SExpr right) {
        if (isInt(left, right)) return imultiply(left, right);
        if (isBv(left, right)) return bvmultiply(left, right);
        return typeException();
    }

    public SExpr divide(SExpr left, SExpr right, boolean b) {
        if (isInt(left, right)) return idivide(left, right);
        if (isBv(left, right)) return bvdivide(left, right);
        return typeException();
    }


    //endregion

    //region integer operations


    @SneakyThrows
    SAtom intValue(String svalue) {
        return symbolAndValueCache.get(svalue, () -> new SAtom(SmtType.INT, ResolvedPrimitiveType.INT, svalue));
    }

    @SneakyThrows
    SAtom intValue(long value) {
        return intValue("" + value);
    }

    @SneakyThrows
    SAtom intValue(BigInteger value) {
        return intValue("" + value);
    }

    public SExpr iadd(SExpr left, SExpr right) {
        return fnApplyToInt("+", left, right);
    }

    public SExpr ilessThan(SExpr left, SExpr right) {
        return fnApplyToBool("<", left, right);
    }

    public SExpr ilessOrEquals(SExpr left, SExpr right) {
        return fnApplyToBool("<=", left, right);
    }

    public SExpr igreaterThan(SExpr left, SExpr right) {
        return fnApplyToBool(">", left, right);
    }

    public SExpr igreaterOrEquals(SExpr left, SExpr right) {
        return fnApplyToBool(">=", left, right);
    }

    public SExpr isubstract(SExpr left, SExpr right) {
        return fnApplyToInt("-", left, right);
    }

    public SExpr imultiply(SExpr left, SExpr right) {
        return fnApplyToInt("*", left, right);
    }

    public SExpr intType() {
        return symbol("Int");
    }

    public SExpr imodulo(SExpr left, SExpr right) {
        return fnApplyToInt("mod", left, right);
    }

    public SExpr idivide(SExpr left, SExpr right) {
        return fnApplyToInt("/", left, right);
    }

    //endregion

    //region bit vectors
    public SExpr bvor(SExpr left, SExpr right) {
        return fnApplyToBV("bvor", left, right);
    }

    public SExpr bvnot(SExpr expr) {
        return fnApplyToBV("bvnot", expr);
    }

    public SExpr bvnegate(SExpr sexpr) {
        return fnApplyToBV("bvneg", sexpr);
    }

    public SExpr bvlessThan(SExpr left, SExpr right) {
        return fnApplyToBool("bvslt", left, right);
    }

    public SExpr bvlessOrEquals(SExpr left, SExpr right) {
        return fnApplyToBool("bvsle", left, right);
    }

    public SExpr bvgreaterThan(SExpr left, SExpr right) {
        return fnApplyToBool("bvsgt", left, right);
    }

    public SExpr bvgreaterOrEquals(SExpr left, SExpr right) {
        return fnApplyToBool("bvsge", left, right);
    }

    public SExpr bvshiftRight(SExpr left, SExpr right, boolean sign) {
        return fnApplyToBV(sign ? "bvashr" : "bvshr", left, right);
    }

    public SExpr bvshiftLeft(SExpr left, SExpr right) {
        return fnApplyToBV("bvshl", left, right);
    }

    public SExpr bvand(SExpr left, SExpr right) {
        return fnApplyToBV("bvand", left, right);
    }

    public SExpr bvadd(SExpr left, SExpr right) {
        return fnApplyToBV("bvadd", left, right);

    }

    private SExpr bvsubstract(SExpr left, SExpr right) {
        return fnApplyToBV("bvsub", left, right);
    }

    private SExpr bvmultiply(SExpr left, SExpr right) {
        return fnApplyToBV("bvmul", left, right);

    }

    private SExpr bvdivide(SExpr left, SExpr right) {
        return fnApplyToBV("bvsdiv", left, right);
    }

    private SExpr bvmodulo(SExpr left, SExpr right) {
        return fnApplyToBV("bvsrem", left, right);
    }

    private SExpr fnApplyToBV(String symbol, SExpr left) {
        return fnApply(null, left.getSmtType(), symbol, left);
    }


    public SExpr makeBitvector(int width, long value) {
        SmtType.BitVec type = SmtType.getBitVec(width);
        return new SList(type, null, symbol("_"), symbol("bv" + width), intValue(value));
    }


    public SExpr makeBitvector(int width, BigInteger value) {
        SmtType.BitVec type = SmtType.getBitVec(width);
        return new SList(type, null, symbol("_"), symbol("bv" + width), intValue(value));
    }


    public SExpr bvType(int width) {
        return new SList(SmtType.TYPE, null, symbol("_"), symbol("BitVec"), intValue(width));
    }
    //endregion

    private boolean isBool(SExpr sexpr) {
        return sexpr.getSmtType() == SmtType.BOOL;
    }

    private boolean isBool(SExpr left, SExpr right) {
        return left.getSmtType() == right.getSmtType() && isBool(left);
    }

    private boolean isBv(SExpr left, SExpr right) {
        return left.getSmtType() == right.getSmtType() && isBv(left);
    }

    private boolean isBv(SExpr left) {
        return left.getSmtType() instanceof SmtType.BitVec;
    }

    private boolean isInt(SExpr left, SExpr right) {
        return left.getSmtType() == right.getSmtType() && isInt(left);
    }

    private boolean isInt(SExpr left) {
        return left.getSmtType() == SmtType.INT;
    }

    private SExpr typeException() {
        throw new RuntimeException("Type mismatch!");
    }


    private SExpr fnApplyToInt(String symbol, SExpr left, SExpr right) {
        return fnApply(ResolvedPrimitiveType.INT, SmtType.INT, symbol, left, right);
    }

    private SExpr fnApplyToBV(String symbol, SExpr left, SExpr right) {
        return fnApply(null, left.getSmtType(), symbol, left, right);
    }


    public SExpr fpType(int width) {
        return new SList(SmtType.TYPE, null, symbol("_"), symbol("FloatingPoint"), intValue(width));
    }

    public SExpr arrayType(SExpr from, SExpr to) {
        return new SList(SmtType.TYPE, null, symbol("Array"), from, to);
    }


    public SExpr list(ResolvedType javaType, SmtType stype, Object... args) {
        SExpr[] nargs = new SExpr[args.length];
        for (int i = 0; i < args.length; i++) {
            Object arg = args[i];
            if (arg instanceof SExpr) nargs[i] = (SExpr) arg;
            else if (arg instanceof String) nargs[i] = symbol((String) arg);
            else typeException("Unhandled type %s", arg.getClass());
        }
        return new SList(stype, javaType, nargs);
    }

    public SExpr makeTrue() {
        return makeBoolean(true);
    }


    public SExpr makeFalse() {
        return makeBoolean(false);
    }

    @SneakyThrows
    public SExpr makeBoolean(boolean value) {
        String v = value ? "true" : "false";
        return symbolAndValueCache.get(v, () -> new SAtom(SmtType.BOOL, ResolvedPrimitiveType.BOOLEAN, v));
    }

    public SExpr makeInt(String value) {
        return intValue(value);
    }

    public SExpr makeNull() {
        return symbol("null");
    }

    public SExpr makeThis() {
        return symbol("this");
    }

    public SExpr makeSuper() {
        return symbol("super");
    }

    public SExpr boolType() {
        return symbol("Bool");
    }

    public SExpr javaObjectType() {
        return symbol("Object");
    }

    public SExpr type(SmtType type) {
        if (type == SmtType.JAVA_OBJECT) return javaObjectType();
        if (type == SmtType.INT) return intType();
        if (type == SmtType.REAL) return realType();
        if (type == SmtType.FP32) return fpType(32);
        if (type == SmtType.FP64) return fpType(64);
        if (type == SmtType.BOOL) return boolType();
        if (type instanceof SmtType.Array)
            return arrayType(type(((SmtType.Array) type).from), type(((SmtType.Array) type).to));
        if (type instanceof SmtType.BitVec) return bvType(((SmtType.BitVec) type).width);

        return typeException();
    }


    private SExpr realType() {
        return symbol("Int");
    }

    public SExpr var(SmtType type, ResolvedType javaType, String name) {
        //nocache, conflict in type could be created
        return new SAtom(type, javaType, name);
    }

    @SneakyThrows
    public SExpr command(String symbol, SExpr... args) {
        return fnApply(null, SmtType.COMMAND, symbol, args);
    }

    public SExpr select(SmtType stype, ResolvedType javaType, SExpr array, SExpr index) {
        return list(javaType, stype, symbol("select"), array, index);
    }

    public SExpr store(SExpr array, SExpr index, SExpr value) {
        return list(array.getJavaType(), array.getSmtType(), symbol("store"), array, index, value);
    }

    public SExpr fieldAccess(ResolvedType javaType, SmtType stype, String field, SExpr obj) {
        return fnApply(javaType, stype, field, obj);
    }

    private SExpr typeException(String fmt, Object... args) {
        throw new RuntimeException(String.format(fmt, args));
    }

    public SExpr let(List<SExpr> vars, SExpr body) {
        return list(body.getJavaType(), body.getSmtType(), "let", list(vars), body);
    }

    public SExpr nonNull(SExpr expr) {
        return not(equality(expr, makeNull()));
    }

    public SExpr binder(SmtType type, String name) {
        return list(null, null, name, type(type));
    }

}

File: tools/smt/src/main/kotlin/io/github/jmltoolkit/smt/SmtTermFactory.kt
package io.github.jmltoolkit.smt

import com.github.javaparser.resolution.types.ResolvedPrimitiveType.BOOLEAN
import com.github.javaparser.resolution.types.ResolvedPrimitiveType.INT
import com.github.javaparser.resolution.types.ResolvedType
import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import io.github.jmltoolkit.smt.model.SAtom
import io.github.jmltoolkit.smt.model.SExpr
import io.github.jmltoolkit.smt.model.SList
import io.github.jmltoolkit.smt.model.SmtType
import io.github.jmltoolkit.smt.model.SmtType.BitVec
import java.math.BigInteger

/**
 * @author Alexander Weigl
 * @version 1 (07.08.22)
 */
object SmtTermFactory {
    private val symbolAndValueCache: Cache<String, SAtom> = CacheBuilder.newBuilder().softValues().build()

    //region boolean operators
    fun and(vararg terms: SExpr): SExpr = and(terms.toList())
    fun and(seq: List<SExpr>): SExpr = fnApply(BOOLEAN, SmtType.BOOL, "and", seq)

    fun or(vararg terms: SExpr): SExpr = or(terms.toList())
    fun or(terms: List<SExpr>) = fnApply(BOOLEAN, SmtType.BOOL, "or", terms)

    fun impl(premise: SExpr, concl: SExpr): SExpr = fnApply(BOOLEAN, SmtType.BOOL, "=>", premise, concl)

    //endregion
    fun ite(cond: SExpr, then: SExpr, otherwise: SExpr): SExpr =
        fnApply(then.javaType!!, then.smtType!!, "ite", cond, then, otherwise)

    fun fnApply(javaType: ResolvedType?, smtType: SmtType, fn: String, arg: SExpr): SExpr =
        SList(smtType, javaType, listOf(symbol(fn), arg))

    private fun fnApply(javaType: ResolvedType?, smtType: SmtType, fn: String, arg1: SExpr, arg2: SExpr): SExpr =
        SList(smtType, javaType, listOf(symbol(fn), arg1, arg2))

    private fun fnApply(
        javaType: ResolvedType?,
        smtType: SmtType,
        fn: String,
        arg1: SExpr,
        arg2: SExpr,
        arg3: SExpr
    ): SExpr = SList(smtType, javaType, listOf(symbol(fn), arg1, arg2, arg3))

    private fun fnApply(javaType: ResolvedType?, smtType: SmtType, fn: String, args: List<SExpr>): SExpr {
        val nargs = mutableListOf<SExpr>(symbol(fn))
        nargs.addAll(args)
        return SList(smtType, javaType, nargs)
    }

    fun symbol(fn: String): SAtom = symbolAndValueCache.get(fn) { SAtom(SmtType.SYMBOL, null, fn) }

    fun forall(variables: List<SExpr>, formula: SExpr): SExpr =
        fnApply(BOOLEAN, SmtType.BOOL, "forall", list(variables), formula)

    fun exists(variables: List<SExpr>, formula: SExpr): SExpr =
        fnApply(BOOLEAN, SmtType.BOOL, "exists", list(variables), formula)


    fun list(javaType: ResolvedType, smtType: SmtType, symbol: SAtom, args: Array<SExpr>): SExpr {
        val nargs = mutableListOf<SExpr>()
        nargs.add(symbol)
        nargs.addAll(args)
        return SList(smtType, javaType, nargs)
    }

    //TODO weigl correct?
    fun list(variables: List<SExpr>): SExpr {
        return SList(null, null, listOf())
    }

    //region polymorphic operators
    fun bor(left: SExpr, right: SExpr): SExpr {
        if (isBool(left, right)) return or(left, right)
        if (isBv(left, right)) return bvor(left, right)
        return typeException()
    }


    fun add(left: SExpr, right: SExpr): SExpr {
        if (isInt(left, right)) return iadd(left, right)
        if (isBv(left, right)) return bvadd(left, right)
        return typeException()
    }

    fun subtract(left: SExpr, right: SExpr): SExpr {
        if (isInt(left, right)) return isubstract(left, right)
        if (isBv(left, right)) return bvsubstract(left, right)
        return typeException()
    }

    fun modulo(left: SExpr, right: SExpr, b: Boolean): SExpr {
        if (isInt(left, right)) return imodulo(left, right)
        if (isBv(left, right)) return bvmodulo(left, right)
        return typeException()
    }

    fun shiftLeft(left: SExpr, right: SExpr): SExpr {
        if (isBv(left, right)) return bvshiftLeft(left, right)
        return typeException()
    }


    fun shiftRight(left: SExpr, right: SExpr, sign: Boolean): SExpr {
        if (isBv(left, right)) return bvshiftRight(left, right, sign)
        return typeException()
    }

    fun lessOrEquals(left: SExpr, right: SExpr, sign: Boolean): SExpr {
        if (isInt(left, right)) return ilessOrEquals(left, right)
        if (isBv(left, right)) return bvlessOrEquals(left, right)
        return typeException("Could not handle types '%s <= %s'", left.smtType, right.smtType)
    }

    fun greaterOrEquals(left: SExpr, right: SExpr, b: Boolean): SExpr {
        if (isInt(left, right)) return igreaterOrEquals(left, right)
        if (isBv(left, right)) return bvgreaterOrEquals(left, right)
        return typeException()
    }

    fun lessThan(left: SExpr, right: SExpr): SExpr {
        if (isInt(left, right)) return ilessThan(left, right)
        if (isBv(left, right)) return bvlessThan(left, right)
        return typeException("Could not handle types '%s <%s'", left.smtType!!, right.smtType!!)
    }

    fun equiv(left: SExpr, right: SExpr): SExpr = fnApply(BOOLEAN, SmtType.BOOL, "=", left, right)

    fun not(expr: SExpr): SExpr {
        if (isBv(expr)) return bvnot(expr)
        if (isBool(expr)) return fnApply(BOOLEAN, SmtType.BOOL, "not", expr)
        return typeException()
    }


    fun xor(left: SExpr, right: SExpr): SExpr = fnApply(BOOLEAN, SmtType.BOOL, "xor", left, right)

    fun equality(left: SExpr, right: SExpr): SExpr = fnApply(BOOLEAN, SmtType.BOOL, "=", left, right)

    fun band(left: SExpr, right: SExpr): SExpr {
        if (isBool(left, right)) return and(left, right)

        if (isBv(left, right)) return bvand(left, right)

        return typeException()
    }


    private fun fnApplyToBool(symbol: String, left: SExpr, right: SExpr): SExpr =
        fnApply(BOOLEAN, SmtType.BOOL, symbol, left, right)

    fun greaterThan(left: SExpr, right: SExpr): SExpr {
        if (isInt(left, right)) return igreaterThan(left, right)
        if (isBv(left, right)) return bvgreaterThan(left, right)
        return typeException("Could not handle types '%s > %s'", left.smtType, right.smtType)
    }


    fun negate(sexpr: SExpr): SExpr {
        if (isBool(sexpr)) return not(sexpr)
        if (isBv(sexpr)) return bvnegate(sexpr)
        return typeException()
    }


    fun multiply(left: SExpr, right: SExpr): SExpr {
        if (isInt(left, right)) return imultiply(left, right)
        if (isBv(left, right)) return bvmultiply(left, right)
        return typeException()
    }

    fun divide(left: SExpr, right: SExpr, b: Boolean): SExpr {
        if (isInt(left, right)) return idivide(left, right)
        if (isBv(left, right)) return bvdivide(left, right)
        return typeException()
    }


    //endregion
    //region integer operations
    fun intValue(svalue: String): SAtom = symbolAndValueCache.get(svalue) { SAtom(SmtType.INT, INT, svalue) }

    fun intValue(value: Long): SAtom = intValue("" + value)

    fun intValue(value: BigInteger): SAtom = intValue("" + value)

    fun iadd(left: SExpr, right: SExpr): SExpr = fnApplyToInt("+", left, right)

    fun ilessThan(left: SExpr, right: SExpr): SExpr = fnApplyToBool("<", left, right)

    fun ilessOrEquals(left: SExpr, right: SExpr): SExpr = fnApplyToBool("<=", left, right)

    fun igreaterThan(left: SExpr, right: SExpr): SExpr = fnApplyToBool(">", left, right)

    fun igreaterOrEquals(left: SExpr, right: SExpr): SExpr = fnApplyToBool(">=", left, right)

    fun isubstract(left: SExpr, right: SExpr): SExpr = fnApplyToInt("-", left, right)

    fun imultiply(left: SExpr, right: SExpr): SExpr = fnApplyToInt("*", left, right)

    fun intType(): SExpr = symbol("Int")

    fun imodulo(left: SExpr, right: SExpr): SExpr = fnApplyToInt("mod", left, right)

    fun idivide(left: SExpr, right: SExpr): SExpr = fnApplyToInt("/", left, right)

    //endregion
    //region bit vectors
    fun bvor(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvor", left, right)

    fun bvnot(expr: SExpr): SExpr = fnApplyToBV("bvnot", expr)

    fun bvnegate(sexpr: SExpr): SExpr = fnApplyToBV("bvneg", sexpr)

    fun bvlessThan(left: SExpr, right: SExpr): SExpr = fnApplyToBool("bvslt", left, right)

    fun bvlessOrEquals(left: SExpr, right: SExpr): SExpr = fnApplyToBool("bvsle", left, right)

    fun bvgreaterThan(left: SExpr, right: SExpr): SExpr = fnApplyToBool("bvsgt", left, right)

    fun bvgreaterOrEquals(left: SExpr, right: SExpr): SExpr = fnApplyToBool("bvsge", left, right)

    fun bvshiftRight(left: SExpr, right: SExpr, sign: Boolean): SExpr =
        fnApplyToBV(if (sign) "bvashr" else "bvshr", left, right)

    fun bvshiftLeft(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvshl", left, right)

    fun bvand(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvand", left, right)

    fun bvadd(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvadd", left, right)

    private fun bvsubstract(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvsub", left, right)

    private fun bvmultiply(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvmul", left, right)

    private fun bvdivide(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvsdiv", left, right)

    private fun bvmodulo(left: SExpr, right: SExpr): SExpr = fnApplyToBV("bvsrem", left, right)

    private fun fnApplyToBV(symbol: String, left: SExpr): SExpr = fnApply(null, left.smtType!!, symbol, left)


    fun makeBitvector(width: Int, value: Long): SExpr = makeBitvector(width, BigInteger.valueOf(value))


    fun makeBitvector(width: Int, value: BigInteger): SExpr {
        val type = SmtType.getBitVec(width)
        return SList(
            type, null,
            listOf(
                SList(null, null, listOf(symbol("_"), symbol("int2bv"), symbol("" + width))),
                intValue(value)
            )
        )
    }


    fun bvType(width: Int): SExpr =
        SList(SmtType.TYPE, null, listOf(symbol("_"), symbol("BitVec"), intValue(width.toLong())))

    //endregion
    private fun isBool(sexpr: SExpr): Boolean = sexpr.smtType == SmtType.BOOL

    private fun isBool(left: SExpr, right: SExpr): Boolean = left.smtType === right.smtType && isBool(left)

    private fun isBv(left: SExpr, right: SExpr): Boolean = left.smtType === right.smtType && isBv(left)

    private fun isBv(left: SExpr): Boolean = left.smtType is BitVec

    private fun isInt(left: SExpr, right: SExpr): Boolean = left.smtType === right.smtType && isInt(left)

    private fun isInt(left: SExpr): Boolean = left.smtType === SmtType.INT

    private fun typeException(): SExpr {
        throw RuntimeException("Type mismatch!")
    }


    private fun fnApplyToInt(symbol: String, left: SExpr, right: SExpr): SExpr =
        fnApply(INT, SmtType.INT, symbol, left, right)

    private fun fnApplyToBV(symbol: String, left: SExpr, right: SExpr): SExpr =
        fnApply(null, left.smtType!!, symbol, left, right)


    fun fpType(width: Int): SExpr =
        SList(SmtType.TYPE, null, listOf(symbol("_"), symbol("FloatingPoint"), intValue(width.toLong())))

    fun arrayType(from: SExpr, to: SExpr): SExpr = SList(SmtType.TYPE, null, listOf(symbol("Array"), from, to))


    fun list(javaType: ResolvedType?, stype: SmtType, vararg args: Any): SExpr {
        val nargs = mutableListOf<SExpr>()
        for (i in args.indices) {
            val arg = args[i]
            nargs.add(
                when (arg) {
                    is SExpr -> arg
                    is String -> symbol(arg)
                    else -> typeException("Unhandled type %s", arg.javaClass)
                }
            )
        }
        return SList(stype, javaType, nargs)
    }

    fun makeTrue(): SExpr = makeBoolean(true)


    fun makeFalse(): SExpr = makeBoolean(false)

    fun makeBoolean(value: Boolean): SExpr {
        val v = if (value) "true" else "false"
        return symbolAndValueCache.get(v) { SAtom(SmtType.BOOL, BOOLEAN, v) }
    }

    fun makeInt(value: String): SExpr = intValue(value)

    fun makeNull(): SExpr = symbol("null")

    fun makeThis(): SExpr = symbol("this")

    fun makeSuper(): SExpr = symbol("super")

    fun boolType(): SExpr = symbol("Bool")

    fun javaObjectType(): SExpr = symbol("Object")

    fun type(type: SmtType): SExpr {
        if (type === SmtType.JAVA_OBJECT) return javaObjectType()
        if (type === SmtType.INT) return intType()
        if (type === SmtType.REAL) return realType()
        if (type === SmtType.FP32) return fpType(32)
        if (type === SmtType.FP64) return fpType(64)
        if (type === SmtType.BOOL) return boolType()
        if (type is SmtType.Array) return arrayType(
            type(type.from),
            type(type.to)
        )
        if (type is BitVec) return bvType(type.width)

        return typeException()
    }


    private fun realType(): SExpr = symbol("Int")

    fun variable(type: SmtType, javaType: ResolvedType?, name: String): SExpr =
        //nocache, conflict in type could be created
        SAtom(type, javaType, name)

    fun command(symbol: String, vararg args: SExpr): SExpr =
        fnApply(null, SmtType.COMMAND, symbol, args.toList())

    fun select(stype: SmtType, javaType: ResolvedType?, array: SExpr, index: SExpr): SExpr =
        list(javaType, stype, symbol("select"), array, index)

    fun store(array: SExpr, index: SExpr, value: SExpr): SExpr =
        list(array.javaType, array.smtType!!, symbol("store"), array, index, value)

    fun fieldAccess(javaType: ResolvedType?, stype: SmtType, field: String, obj: SExpr): SExpr =
        fnApply(javaType, stype, field, obj)

    private fun typeException(fmt: String, vararg args: Any?): SExpr {
        throw RuntimeException(String.format(fmt, *args))
    }

    fun let(vars: List<SExpr>, body: SExpr): SExpr = list(body.javaType, body.smtType!!, "let", list(vars), body)

    fun nonNull(expr: SExpr): SExpr = not(equality(expr, makeNull()))

    fun binder(type: SmtType, name: String): SExpr = list(null, SmtType.JAVA_OBJECT, name, type(type))
}
